metaserver/src/main/scala/scala/meta/languageserver/OutlineProvider.scala

```diff
@@ -0,0 +1,64 @@
+package scala.meta.languageserver
+
+import scala.collection.mutable
+import com.typesafe.scalalogging.LazyLogging
+import org.langmeta.io.AbsolutePath
+import langserver.{types => l}
+import langserver.messages.DefinitionResult
+import scala.meta.languageserver.ScalametaEnrichments._
+import scala.meta._
+
+object OutlineProvider extends LazyLogging {
```

Comment:
Shouldn't this be called `DocumentSymbolProvider`?

Reply:
> What do you think about moving all *Provider classes to the providers package? We can do that in a separate PR, in which case it's fine to keep OutlineProvider here for now.

I think it will be an improvement. But In general it's not clear to me what a provider is supposed to be: a thing to implement related features with common context or just a per-feature implementation. I would vote for the former. 

> Shouldn't this be called DocumentSymbolProvider?

For this reason I did it call it this way. I thought it could also provide `workspace/symbol` functionality in the future. It could be called `SymbolProvider` then, but I think it's a bit confusing.

Reply:
> I think it will be an improvement. But In general it's not clear to me what a provider is supposed to be: a thing to implement related features with common context or just a per-feature implementation. I would vote for the former.

I vote for one *Provider per LSP message, so `DocumentSymbolProvider` and `WorkspaceSymbolProvider`. We might actually want to highlight different kinds of symbols in workspaces, for example prioritize global class/object/trait.

Reply:
OK, I will rename it to `DocumentSymbolProvider` for now. 
But in general, what about context sharing? For example here, we could maintain a `Map[AbsolutePath, List[SymbolInformation]]` instead of `Map[AbsolutePath, Source]` to avoid traversal in case the buffer doesn't parse. This is something that could be reused for document or workspace symbols outline. Same with `ReferenceProvider` and `DefinitionProvider` sharing `SymbolIndex`.

Reply:
> But in general, what about context sharing?

I think context/state should be isolated to ScalametaLanguageServer as much as possible. The *Provider objects should contain only simple stateless methods, so those should accept the dependencies (Source/List[SymbolInformation]/..) as arguments. I'm still learning about Monix but I feel like we should be able to hide away a lot of the state with Observable[T], just need more time to wrap my head around it.

Reply:
Renamed in 8712ca28169ffa181df2f08cc0b1d71aefadec14

Reply:
> I think context/state should be isolated to ScalametaLanguageServer as much as possible. The *Provider objects should contain only simple stateless methods, so those should accept the dependencies (Source/List[SymbolInformation]/..) as arguments.

Missed this comment. I see your point 👍 Let's keep it simple now then and see how it can be improved with Monix.